x-pack/plugins/monitoring/server/plugin.ts

```diff
@@ -0,0 +1,375 @@
+  private config?: MonitoringConfig;
+  private needToSetupLegacy: boolean = true;
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.initializerContext = initializerContext;
+    this.log = initializerContext.logger.get(LOGGING_TAG);
+    this.getLogger = (...scopes: string[]) => initializerContext.logger.get(LOGGING_TAG, ...scopes);
+  }
+
+  async setup(core: CoreSetup, plugins: PluginsSetup) {
+    this.coreSetup = core;
+    this.setupPlugins = plugins;
+    const config = await this.initializerContext.config
+      .create<MonitoringConfig>()
+      .pipe(first())
+      .toPromise();
+
+    this.config = config;
+    // Monitoring creates and maintains a connection to a potentially
+    // separate ES cluster - create this first
+    const elasticsearchConfig = new ElasticsearchConfig(
+      config.ui.elasticsearch as ElasticsearchConfigType
+    );
+    const cluster = (this.cluster = instantiateClient({
+      log: this.log,
+      elasticsearchConfig,
+      elasticsearchPlugin: {
+        createCluster: core.elasticsearch.createClient,
+      },
+    }));
+
+    // Start our license service which will ensure
+    // the appropriate licenses are present
+    this.licenseService = new LicenseService().setup({
+      licensing: plugins.licensing,
+      monitoringClient: cluster,
+      config,
+      log: this.log,
+    });
+    await this.licenseService.refresh();
+
+    if (KIBANA_ALERTING_ENABLED && this.cluster && this.getLogger) {
+      plugins.alerting.registerType(
+        getLicenseExpiration(
+          () => this.uiSettingsService,
+          this.cluster,
+          this.getLogger,
+          config.ui.ccs.enabled
+        )
+      );
+    }
+
+    if (config.ui.enabled) {
+      this.registerPluginInUI(plugins);
+    }
+
+    return {
+      registerLegacyAPI: (legacyAPI: LegacyAPI) => {
+        this.legacyAPI = legacyAPI;
+        if (this.needToSetupLegacy) {
+          this.setupLegacy();
+        }
+      },
+    };
+  }
+
+  async start(core: CoreStart, plugins: PluginsStart) {
+    if (!this.coreSetup || !this.licenseService || !this.cluster) {
+      return;
+    }
+
+    const [config, legacyConfig] = await combineLatest([
+      this.initializerContext.config.create<MonitoringConfig>(),
+      this.initializerContext.config.legacy.globalConfig$,
+    ])
+      .pipe(first())
+      .toPromise();
+
+    // Create our shim which is currently used to power our routing
+    this.monitoringCore = await this.getLegacyShim(
+      config,
+      legacyConfig,
+      this.coreSetup,
+      plugins,
+      this.licenseService,
+      this.cluster
+    );
+
+    if (config.ui.enabled) {
+      requireUIRoutes(this.monitoringCore);
+      initInfraSource(this.monitoringCore.config(), this.setupPlugins?.infra);
+    }
+
+    this.uiSettingsService = core.uiSettings;
+    if (this.legacyAPI) {
+      this.setupLegacy();
+      this.needToSetupLegacy = false;
+    }
+  }
+
+  stop() {
```

Comment:
I believe you need to stop licensing polling as well

Reply:
Yes, good call